### ISSUE TITLE
create a stream from a promise

### ISSUE THREAD

**Issue by catdad** [stezu/node-stream#23]
So that I can asynchronously create a stream from stuff that is not a stream.

Pseudo-code:

```javascript
ns.createStream.fromPromise(globby('**/*'));
```

This could just resolve the promise an internally use `ns.createStream(...args)` proposed in #22, since some promises insist on resolving with more than one arg.

**Comment by stezu** [comment]
@catdad it seems like you have the opinion that a promise which resolves with multiple arguments should result in a stream with the same number of items. Is that accurate?

And either way this is a completely new implementation and can't use `fromArray` since we have to return a stream immediately. So, the implementation would be similar to `fromArray` in that we'll be creating a readable stream immediately and return it, but the `_read` function will need to wait for the promise to resolve before writing data to the stream. And when it does, we'll be writing each argument on the resolve as an individual item. Right?

**Comment by stezu** [comment]
Okay, so the native promise implementation does not support resolving with multiple arguments and it's not part of the spec. Apparently, Bluebird, Q, Chrome and Node (and probably others) don't implement the multi-arg promise resolution.

@catdad where did you see a multi-arg promise?

**Comment by catdad** [comment]
@stezu It is 100% part of the spec: https://promisesaplus.com/#point-23

And here is some code:

```javascript
function rejected() {
  return new Promise(function (resolve, reject) {
    setImmediate(reject, new Error('pineapples'));
  });
}

rejected().then(function () {
  console.log('great success');
}, function (err) {
  console.log('error:', err);
});
```

And here's me running it in Node 6:

![image](https://cloud.githubusercontent.com/assets/2205537/25509328/0d0c1ca8-2b86-11e7-89af-8b0312e941b6.png)

**Comment by catdad** [comment]
All that the promise spec says is that `onFulfilled` must be called with the first argument as its value... which I strictly interpret as needing to have only one argument. But in the wild, I have seen promises fulfill with more than one argument and it works fine. Which makes me wonder, what does `async/await` do in this case?

**Comment by catdad** [comment]
Just checked native Node 6 promises and bluebird, and neither of those allowed me to use more than one value.

```javascript
function sucessWithArgs() {
  return new Promise(function (resolve, reject) {
    setImmediate(resolve, 1, 2, 3, 4);
  });
}

sucessWithArgs().then(function () {
  console.log('great success', arguments);
}, function (err) {
  console.log('error:', err);
});
```

Logged:

```
C:\Users\Kiril\Desktop\New folder (2)>node promise-test.js
great success { '0': 1 }
```

So yeah... what I have seen are probably crappy promise implementations, and it seems that it would be okay to only care about the first value that the promise resolves with... which would only matter in the case where we are given one of those crappy promises.

**Comment by catdad** [comment]
Just so you don't say I didn't complete the investigation, here's `async/await`:

```javascript
function sucessWithArgs() {
  return new Promise(function (resolve, reject) {
    setImmediate(resolve, 1, 2, 3, 4);
  });
}

async function thing() {
  console.log(await sucessWithArgs());
}

thing();
```

![image](https://cloud.githubusercontent.com/assets/2205537/25509632/67b18aa6-2b88-11e7-8d31-051290a01156.png)

**Comment by catdad** [comment]
Also, I don't appreciate going from the guy who hates promises to being a promise expert. You need to fix this.

**Comment by stezu** [comment]
@catdad I think I lost the thread. What are you saying?

**Comment by catdad** [comment]
**tl;dr**

* Seems like all the standard things do not allow you to resolve a promise with more than one arg. If you do, it's dropped anyway. I don't feel strongly about supporting poorly-implemented promises.
* `.then(onFulfilled, onRejected)` is definitely a thing and it is supported everywhere, so I don't know what you are talking about.

**Comment by stezu** [comment]
I wasn't saying `.then(onFulfilled, onRejected)` wasn't a thing, I was saying `resolve('banana', 'apple')` was not a thing. So, we're on the same page then. Cool, so I implemented the thing in a branch. I want to implement `fromCallback` before pushing 1.6.0 and then we can get to crazy things like generators, sets and iterables.

**Comment by catdad** [comment]
👍 

Ooh, generators.
